src/ball.rs

```diff
@@ -0,0 +1,121 @@
+//! Balls, including circles and spheres.
+
+use crate::{
+    aabb::{Aabb, IntersectsAabb},
+    bounding_hierarchy::BHValue,
+};
+use nalgebra::Point;
+
+/// A 2d circle.
+pub type Circle<T> = Ball<T, 2>;
+
+/// A 3d sphere.
+pub type Sphere<T> = Ball<T, 3>;
+
+/// In 2D, a circle. In 3D, a sphere. This can be used for traversing BVH's.
+#[derive(Debug, Clone, Copy)]
+pub struct Ball<T: BHValue, const D: usize> {
+    /// The center of the ball.
+    pub center: Point<T, D>,
+    /// The radius of the ball.
+    pub radius: T,
+}
+
+impl<T: BHValue, const D: usize> Ball<T, D> {
+    /// Creates a [`Ball`] with the given `center` and `radius`.
+    ///
+    /// # Panics
+    /// Panics, in debug mode, if the radius is negative.
+    ///
+    /// # Examples
+    /// ```
+    /// use bvh::ball::Ball;
+    /// use nalgebra::Point3;
+    ///
+    /// let ball = Ball::new(Point3::new(1.0, 1.0, 1.0), 1.0);
+    /// assert_eq!(ball.center, Point3::new(1.0, 1.0, 1.0));
+    /// assert_eq!(ball.radius, 1.0)
+    /// ```
+    ///
+    /// [`Ball`]: struct.Ball.html
+    pub fn new(center: Point<T, D>, radius: T) -> Self {
+        debug_assert!(radius >= T::from_f32(0.0).unwrap());
+        Self { center, radius }
+    }
+
+    /// Returns true if this [`Ball`] contains the [`Point`].
+    ///
+    /// # Examples
+    /// ```
+    /// use bvh::ball::Ball;
+    /// use nalgebra::Point3;
+    ///
+    /// let ball = Ball::new(Point3::new(1.0, 1.0, 1.0), 1.0);
+    /// let point = Point3::new(1.25, 1.25, 1.25);
+    ///
+    /// assert!(ball.contains(&point));
+    /// ```
+    ///
+    /// [`Ball`]: struct.Ball.html
+    pub fn contains(&self, point: &Point<T, D>) -> bool {
+        let mut distance_squared = T::zero();
+        for i in 0..D {
+            distance_squared += (point[i] - self.center[i]).powi(2);
+        }
+        distance_squared <= self.radius.powi(2)
```

Comment:
While it doesn't matter much, it might be more intuitive to `sqrt` the `distance` than to `pow2` the original radius. Also in the math below. You can also leave it like this, your call.

Reply:
What I wrote was based on a very common optimization, motivated by `distance_squared.sqrt()` being slower than `self.radius * self.radius` (what `self.radius.powi(2)` is almost certainly optimized to)